bpf_host: emit '-> network' traces for egress packets

ti-mo opened the pull request:
<!-- Description of change -->

Fixes: https://github.com/cilium/cilium/issues/12561

See the issue above for the full context. The caveat is, this program is only attached to the native device egress if either the host firewall or BPF NodePort is enabled, so there still won't be any `TRACE_TO_NETWORK` visibility if that's not the case.

I've had to move the trace to fire before `nodeport_nat_fwd()` is called, since that function tail calls on the happy path, and there's no CT context available downstream from there. This could likely be improved when/if we adopt BTF and get rid of tail calls on kernels that support it.

I've also added `handle_to_netdev_ipv4()` to clean up `to_netdev()` a bit, to bring it on par with the ipv6 handling.

```release-note
bpf_host: emit '-> network' traces for packets leaving the node when Host Firewall or BPF NodePort are enabled
```

ti-mo committed:
bpf_host: declare trace monitor in to_netdev, pass to ipv6 datapath

For https://github.com/cilium/cilium/issues/12561.

To be able to send aggregated traces from 'to_netdev()', the monitor
variable needs to be declared there and passed to the datapath as an
argument. This is the IPv6 path, IPv4 will follow separately as it
moves some logic to a func.

checkpatch.pl trips over srcID, so that was changed to src_id.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf_host: add handle_to_netdev_ipv4, pass monitor to ipv4 datapath

For https://github.com/cilium/cilium/issues/12561.

Moved ipv4 egress srcid resolution to 'handle_to_netdev_ipv4' for parity
with how ipv6 is handled.

Also plumbs the aggregation monitor through to the ipv4 datapath.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf_host: add TRACE_TO_NETWORK trace when host firewall or BPF NodePort are enabled

For https://github.com/cilium/cilium/issues/12561.

Also adds monitor aggregation to the TRACE_TO_NETWORK observation point,
which has the side effect of enabling it by default.

Signed-off-by: Timo Beckers <timo@isovalent.com>

joestringer commented on bpf/bpf_host.c:
```diff
@@ -997,20 +997,31 @@ int to_netdev(struct __ctx_buff *ctx __maybe_unused)
+	 */
+	send_trace_notify(ctx, TRACE_TO_NETWORK, src_id, 0, 0,
+			  0, ret, monitor);
+
```
I don't think it's a great idea to move this before the NAT, because then the trace notification that is emitted will not represent the final state of the packet when the packet leaves the Cilium datapath. For instance, we will not know how NAT was applied. (Admittedly at the moment we don't trace such messages at all, but still, I don't think this is quite getting to where we want to be around notifications).

Realistically to solve monitor aggregation for the NAT case as well, I think we need to pass the `monitor` value through to the tail call (via `skb->cb[]` (preferred) or `skb->mark`) and then ensure that before that program returns the TC result, we perform tracing using the monitor value.

ti-mo replied:
Okay, I wasn't sure how to implement this to begin with, but this would indeed be a great feature to have from the user's perspective. I'll take a stab at it. :+1:

ti-mo replied:
I've added a naive implementation of this in 0e4192a1af281327f4383332407a18209203c9a4. I have some general questions I'll add as a review to this PR.

joestringer commented on bpf/lib/trace.h:
```diff
@@ -141,6 +141,12 @@ static __always_inline bool emit_trace_notify(__u8 obs_point, __u32 monitor)
+	 * Ignore sample when aggregation is enabled and monitor is 0.
+	 * Trace points that exist outside of CT context cannot do rate limiting,
+	 * so they frequently set monitor to 0 to be ignored when aggregation is
+	 * enabled (the default).
```
nit, I suspect that when you say "trace points" here you are referring to the BPF code in the tree which calls `send_trace_notify()`. However, "trace points" in BPF context has a complete other meaning relating to where in the kernel that the BPF program is attached, which makes the comment ambiguous.

Also, we often use the word "context" in BPF to refer to the parameters passed into a BPF hook, so reusing that word feels a bit imprecise to me.

Here's a couple of alternate suggestions that I think would still be correct and may provide the additional context that will help future readers:

```suggestion
	 * Hook points that do not perform connection tracking cannot do rate limiting,
	 * so they frequently set monitor to 0 to be ignored when aggregation is
	 * enabled (the default).
```

Or with fewer negative statements,

```suggestion
	 * Rate limiting relies on connection tracking, so if the BPF program does
	 * not perform CT then monitor will be set to 0 to avoid emitting trace
	 * notifications when aggregation is enabled (the default).
```

ti-mo replied:
> "trace points" in BPF context

> the word "context" in BPF

Absolutely, that would indeed be confusing.. In lieu of what I mean here by 'trace points', we generally use 'observation points' throughout the codebase, I would consider that to be the most accurate description.

ti-mo replied:
I've made the following change in b189308:

```
+     * Ignore sample when aggregation is enabled and 'monitor' is set to 0.
+     * Rate limiting (trace message aggregation) relies on connection tracking,
+     * so if there is no CT information available at the observation point,
+     * then 'monitor' will be set to 0 to avoid emitting trace notifications
+     * when aggregation is enabled (the default).
```

I've also clarified what is meant by 'rate limiting'. WDYT? @joestringer

joestringer replied:
:ok_hand: perfect.

joestringer reviewed (changes requested):
Nice, I think this is quite close to what I had in mind.

Some feedback below on the main open topic, where to execute the trace_notify for these programs.

ti-mo commented on bpf/bpf_host.c:
```diff
@@ -163,6 +163,7 @@ handle_ipv6(struct __ctx_buff *ctx, __u32 secctx, const bool from_host)
 	void *data, *data_end;
 	struct ipv6hdr *ip6;
 	union v6addr *dst;
+	__u32 __maybe_unused monitor = 0;
```
What is `__maybe_unused` for, and where should it generally be used?

pchaigno replied:
It's used to indicate e.g. the `monitor` variable may be unused in some configurations.
https://github.com/cilium/cilium/blob/1843b4adeddad7ee0fd83bc3c71ba6e4d57db48e/bpf/include/bpf/compiler.h#L16

joestringer replied:
`__maybe_unused` is useful when the variable will be conditionally used depending on other codepaths being enabled in macros. In this case, it looks like it's declared at this level, and only conditionally modified if CT is enabled, then it is discarded. So if CT is disabled via macros then no code interacts with `monitor`. If we didn't mark it as `__maybe_unused`, the compiler would complain about the unused variable in that configuration (should be caught by `make -C bpf build_all`).

ti-mo replied:
Of course, that makes sense. I thought in that case the compiler would mark it as being dead code and just optimize it away, but it's good to have it explicitly marked as such. Thanks!

ti-mo commented on bpf/lib/nodeport.h:
```diff
@@ -188,13 +188,19 @@ static __always_inline int nodeport_nat_ipv6_fwd(struct __ctx_buff *ctx,
+
 	if (ret == NAT_PUNT_TO_STACK)
-		ret = CTX_ACT_OK;
+		return CTX_ACT_OK;
```
I hope this is okay to do. From my understanding, `NAT_PUNT_TO_STACK` is returned from NAT codepaths to indicate that a packet should be processed by another BPF program, but I still have no clue as to how that works or why it is necessary. Where can I learn more about this?

joestringer replied:
`git blame` is your friend. Running git blame on lines of code with `NAT_PUNT_TO_STACK` points towards commits 428f43e177de53fe7da0fd885da2eb577fdb071d and efd1ad80bd4ed20388a968185dfe13f920fd1196 with more context around the introduction of the `NAT_PUNT_TO_STACK` return code.

Though I don't understand why the early return in this case, don't we want to call `send_trace_notify()` even in the case that `snat_v6_process()` returns `NAT_PUNT_TO_STACK`?

ti-mo replied:
> `git blame` is your friend

Thanks, hadn't noticed those other occurrences. `git blame` on the occurrences in `nodeport.h` didn't yield anything particularly useful. To which extent do we consider git history documentation? As a codebase goes through refactors, or when those particular lines get touched in any way, the paper trail starts to fade.

The commit messages do provide the proper context and the reason behind the change, but don't answer the question where packet processing eventually continues after they've been passed to the stack.

> don't we want to call `send_trace_notify()` even in the case that `snat_v6_process()` returns `NAT_PUNT_TO_STACK`?

That's exactly what I can't figure out. Does punting to the stack mean a packet is put on the wire? If that's not a 100% sure, we can't log a `TRACE_TO_NETWORK` here, since further processing might mean the packet gets steered elsewhere, e.g., a tunnel. (I don't know)

ti-mo replied:
I've done some testing with the early return, and it looks like `NAT_PUNT_TO_STACK` get set in the majority of cases, causing the trace to not being emitted. That change has been reverted.

ti-mo commented on bpf/lib/nodeport.h:
```diff
@@ -188,13 +188,19 @@ static __always_inline int nodeport_nat_ipv6_fwd(struct __ctx_buff *ctx,
+		return CTX_ACT_OK;
+
+	monitor = ctx_load_meta(ctx, CB_CT_MONITOR);
+	send_trace_notify(ctx, TRACE_TO_NETWORK, 0, 0, 0, 0, ret, monitor);
```
`send_trace_notify` is often provided a `src_id` argument. How can this be obtained from a function like `nodeport_nat_ipv6_fwd`? I've seen `src_id = ipcache_lookup_srcid{4,6}(...)`, but those lookups look like they're quite lengthy (instruction-wise). Would they be worth including here?

joestringer replied:
I *think* we can follow the same idea of placing it into the skb metadata as what I had suggested with the monitor value, functions like to_netdev() should already have resolved the source ID so it's just a matter of transferring that into the skb during the potential tail call, reading it back out in `tail_handle_nat_fwd_ip*()` and passing it along.

ti-mo reviewed (commented):
Another question I had: I recall seeing a test (suite) that checks for particular trace messages being emitted under certain packet scenarios. Could someone point me in the right direction? Should we add some tests to cover this PR as well?

ti-mo committed:
bpf/lib/common.h: document CB_* enum, remove repetitive comments

Also added `CB_CT_MONITOR` to store monitor aggregation value.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf_host: emit TRACE_TO_NETWORK from BPF NodePort datapath

For https://github.com/cilium/cilium/issues/12561.

Store monitor in skb->cb for use in nodeport tail calls.

Signed-off-by: Timo Beckers <timo@isovalent.com>
